### PR TITLE
Add GpuArrayBuffer and BatchedUniformBuffer

### PR DESCRIPTION
**Pull request by JMS55**
# Objective

- Add a type for uploading a Rust `Vec<T>` to a GPU `array<T>`.
- Makes progress towards https://github.com/bevyengine/bevy/issues/89.

## Solution

- Port @superdump's `BatchedUniformBuffer` to bevy main, as a fallback for WebGL2, which doesn't support storage buffers. 
  - Rather than getting an `array<T>` in a shader, you get an `array<T, N>`, and have to rebind every N elements via dynamic offsets.
- Add `GpuArrayBuffer` to abstract over `StorageBuffer<Vec<T>>`/`BatchedUniformBuffer`.

## Future Work
Add a shader macro kinda thing to abstract over the following automatically: https://github.com/bevyengine/bevy/pull/8204#pullrequestreview-1396911727

---

## Changelog
* Added `GpuArrayBuffer`, `GpuComponentArrayBufferPlugin`, `GpuArrayBufferable`, and `GpuArrayBufferIndex` types.
* Added `DynamicUniformBuffer::new_with_alignment()`.

### CODE REVIEW

File: crates/bevy_render/src/render_resource/batched_uniform_buffer.rs
Comment (robtfm) [inline crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:76]: i think maybe the offset should reset to zero when a uniform gets flushed out? not sure

File: crates/bevy_render/src/render_resource/gpu_list.rs
Comment (robtfm) [inline crates/bevy_render/src/render_resource/gpu_list.rs:37]: this 3 maybe originated in the shader cache `NO_STORAGE_BUFFERS` test, where i think it was chosen because the light bindings needed 3 storage buffers. as written it'll always be true or always false, so that if a card supports 8 (the default) but a pipeline creates 10 `GpuList`s, it'll fall over.

ideally this would use the actual number of storage buffers created so far, so that it could use storage buffers up to the actual limit before falling back to uniforms (and shader code could query appropriately using the same limit). i was thinking about passing an index in (or maybe a `<const I: usize>` generic parameter) to allow users to say what the count so far is for the current pipeline/stage. it's not perfect since the caller would need to know all the GpuList bindings upfront (across view/material/mesh etc, e.g.)... but better than just using 3 i think.

not sure, but i think this needs addressing somehow.

**Review by robtfm — COMMENTED** [review]
code and comment quality is good. the code broadly looks right but i really need an example to kick it around properly. i don't think the feature needs an example in the repo but maybe you have something you've been using while building it that i could look at?

**Comment by JMS55** [comment]
I have some messy code [here](https://github.com/JMS55/bevy/compare/gpu-list...JMS55:bevy:gpu-buffers) that uses GpuList for MeshUniforms. If it's not helpful, I can write up a separate example. The next step after this PR will be to use GpuList for mesh MeshUniforms, and then materials.

**Comment by JMS55** [comment]
Reminder to myself that I need to add robswain to the commit authors.

File: crates/bevy_render/src/render_resource/batched_uniform_buffer.rs
Comment (superdump) [reply crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:76]: No. `current_offset` is the dynamic offset at which to bind the uniform buffer for the current fixed-size array of items. 'Flushing' is about pushing the current full fixed-size array that should be bound at `current_offset`, and starting a new binding at a new dynamic offset. So here we increment `current_offset` by the serialised size of `self.temp` rounded up to the dynamic offset alignment. Does that make sense?

File: crates/bevy_render/src/render_resource/gpu_list.rs
Comment (superdump) [reply crates/bevy_render/src/render_resource/gpu_list.rs:37]: I'm not really sure whether bevy should bother trying to check resource usage (like storage buffer bindings per stage or whatever the limit is) because wgpu anyway checks it and points it out. I suppose we could add some validation into our `RenderDevice` wrapper when creating bind group layouts and/or bind groups. The pros of us also doing validation would be that we could perhaps provide what we think are more useful error messages. The cons... well, it's more code in bevy that has to be maintained. :) Thoughts?

File: crates/bevy_render/src/render_resource/batched_uniform_buffer.rs
Comment (robtfm) [reply crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:76]: yes, thanks

File: crates/bevy_render/src/render_resource/gpu_list.rs
Comment (robtfm) [reply crates/bevy_render/src/render_resource/gpu_list.rs:37]: i'm not saying bevy should track it automatically in any scenario, but pipeline authors could do so.

as it is it pipeline authors using `GpuList` will have to choose a minimum spec to support, and use only that many `GpuList` bindings, and fall back to dynamic uniforms for more. then they would still support platforms with no storage buffers, but not platforms with x storage buffers where 3 <= x < what they use.

i don't know if storage buffers per stage is a limit that varies much or is ever low ... if not then no big deal. if so then it sounds fragile (if author chooses to use a lot) or limiting (if they choose to use fewer than they could).

i guess it doesn't matter much for meshes since you're going to use precisely one storage buffer ... certainly happy to kick it down the road to avoid blocking that.

**Comment by JMS55** [comment]
I kind of fixed the commit history... good enough ig?

File: crates/bevy_render/src/render_resource/gpu_list.rs
Comment (superdump) [reply crates/bevy_render/src/render_resource/gpu_list.rs:37]: It's a good point. And good idea about making it so that one can perhaps prioritise bindings that should be storage buffers, and then overflow to uniform buffers. That said, I think it should be handled in a separate PR as it is probably quite an intrusive change to make and needs more thought to make it work well and conveniently.

**Review by superdump — CHANGES_REQUESTED** [review]
Just a few small changes. Otherwise LGTM. I look forward to using this. :)

One nice to have though I'm not sure how to do it - it would be useful to have a nice shader abstraction for it. I suppose the shader side will be either:
```rust
var<uniform> my_list: array<T, #{T_BATCH_SIZE}>;
```
or
```rust
var<storage> my_list: array<T>;
```
but either way, the array will be indexed into, so the interface should be the same. So we need a way of setting the binding type to either `uniform` or `storage`, and if uniform then setting the array size. Currently I think one could do that like this:
```rust
#ifdef T_BATCH_SIZE
var<uniform> my_list: array<T, #{T_BATCH_SIZE}u>;
#else
var<storage> my_list: array<T>;
#endif
```

File: crates/bevy_render/src/render_resource/batched_uniform_buffer.rs
Comment (teoxoy) [reply crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:120]: Thanks! If this is ready for production I can merge the branch in encase and do a release.
Let me know!

**Comment by JMS55** [comment]
> One nice to have though I'm not sure how to do it - it would be useful to have a nice shader abstraction for it. I suppose the shader side will be either:

Sadly we don't have a macro system (nor am I sure if we'd want one, those are easy to abuse...), so I'm not sure how we'd do that. Maybe some kind of custom thing in our shader parser, like this:

```rust
var<gpu_list> my_list: GpuList<T>;
```

that the parser replaces with the appropriate stuff, before passing it to naga.

File: crates/bevy_render/src/render_resource/gpu_list.rs
Comment (JMS55) [reply crates/bevy_render/src/render_resource/gpu_list.rs:26]: I modified all of the doc comments to be exactly identical, even if it's the current resource. I guess I could undo that. Maybe it would be better to just replace "other options" with "list of options" or something similar?

File: crates/bevy_render/src/render_resource/gpu_list.rs
Comment (JMS55) [reply crates/bevy_render/src/render_resource/gpu_list.rs:124]: I'd rather say "when setting the bind group in a pass". Compute shaders are a thing.

File: crates/bevy_render/src/render_resource/batched_uniform_buffer.rs
Comment (superdump) [reply crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:124]: It works fine for us. :) There is that other aspect of being able to start the next dynamic offset binding of a uniform buffer at the next dynamic offset alignment if not all space is used, and ensure that the final binding is full-size. I don't know if that would clash with this and basically immediately deprecate this approach. If so maybe you'd prefer that we use a solution in bevy for what we need and add the long-term and more flexible solution to encase when someone gets to it. What do you think?

File: crates/bevy_render/src/render_resource/gpu_list.rs
Comment (superdump) [reply crates/bevy_render/src/render_resource/gpu_list.rs:123]: Sounds good to me.

**Comment by superdump** [comment]
JMS55 confirmed on Discord that I can commit the two outstanding proposals and rebase this PR to get it merged as they are busy with other things at the moment.

File: crates/bevy_render/src/render_resource/batched_uniform_buffer.rs
Comment (superdump) [reply crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:124]: I won't block the PR on this. We can figure it out over time. :)

File: crates/bevy_render/src/render_resource/batched_uniform_buffer.rs
Comment (superdump) [inline crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:95]: Someone on Discord tried this on web and hit some issues so I debugged the problems and found that the binding given by `DynamicUniformBuffer<MaxCapacityArray<T>>` is using `MaxCapacityArray<T>::min_size()` as the binding size, but that is wrong as `MaxCapacityArray<T>` is runtime-sized. So we have to make sure to fix up the binding size before returning it. 
```suggestion
    #[inline]
    pub fn binding(&self) -> Option<BindingResource> {
        let mut binding = self.uniforms.binding();
        if let Some(BindingResource::Buffer(binding)) = &mut binding {
            // MaxCapacityArray is runtime-sized so can't use T::min_size()
            binding.size = Some(self.size());
        }
        binding
    }
```

File: crates/bevy_render/src/render_resource/gpu_list.rs
Comment (superdump) [inline crates/bevy_render/src/render_resource/gpu_list.rs:88]: This was the other part of the web problem noted on Discord. At this point we don't have access to the `BatchedUniformBuffer<T>` to know what size it should be. We can either construct one and use its runtime size, or we can let wgpu figure out. The latter felt a little cleaner.
```suggestion
                    // BatchedUniformBuffer uses a MaxCapacityArray that is runtime-sized, so we use
                    // None here and let wgpu figure out the size.
                    min_binding_size: None,
```

File: crates/bevy_render/src/render_resource/batched_uniform_buffer.rs
Comment (superdump) [reply crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:137]: I tried to rebase to give credit on the original commit but due to merges it was a pain. I instead added a comment and a co-authored-by so that when the squash merge is done, the credit will follow along with it.

**Review by superdump — APPROVED** [review]

File: crates/bevy_render/src/render_resource/gpu_list.rs
Comment (james-j-obrien) [reply crates/bevy_render/src/render_resource/gpu_list.rs]: nit: This was only changed in one place, for consistency the rest should also be `== 0`

File: crates/bevy_render/src/render_resource/gpu_list.rs
Comment (james-j-obrien) [inline crates/bevy_render/src/render_resource/gpu_list.rs:84]: This should also be changed to `== 0` for consistency.

As well as on line 112.

File: crates/bevy_render/src/render_resource/gpu_list.rs
Comment (james-j-obrien) [inline crates/bevy_render/src/render_resource/gpu_list.rs:112]: Other instance for `== 0`

**Comment by mockersf** [comment]
> * Added `GpuList`, `GpuComponentListPlugin`, and `GpuListable` types.

Could we think of other names for those? For me, `GpuList` is very much not clear on what it is, first idea would be a list of GPUs...

File: crates/bevy_render/src/render_resource/batched_uniform_buffer.rs
Comment (teoxoy) [reply crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:137]: Ok, we can further iterate and see what we come up with. Thanks for the credit!

**Comment by JMS55** [comment]
Due to popular demand, GpuList -> GpuArrayBuffer. I think this PR is ready to go now :)

File: crates/bevy_render/src/render_resource/batched_uniform_buffer.rs
Comment (IceSentry) [inline crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:36]: Might be worth mentioning somewhere that the data is initially pushed to this buffer. I was a bit confused at why there was a `temp` property here.

File: crates/bevy_render/src/render_resource/gpu_array_buffer.rs
Comment (IceSentry) [inline crates/bevy_render/src/render_resource/gpu_array_buffer.rs:15]: Just to keep naming a bit more consistent.

```suggestion
/// Stores an array of elements to be transferred to the GPU and made accessible to shaders as a read-only array.
```

File: crates/bevy_render/src/render_resource/gpu_array_buffer.rs
Comment (IceSentry) [inline crates/bevy_render/src/render_resource/gpu_array_buffer.rs:84]: Is there a reason why it's not just taking &self and using that to determine the layout?

I guess it's just more flexible this way.

**Review by IceSentry — APPROVED** [review]
couple of tiny nits, but otherwise LGTM

File: crates/bevy_render/src/render_resource/gpu_array_buffer.rs
Comment (JMS55) [reply crates/bevy_render/src/render_resource/gpu_array_buffer.rs:84]: It's easier and worked better in practice to not have to query the GpuArrayBuffer to create the BGL for it.

File: crates/bevy_render/src/render_resource/batched_uniform_buffer.rs
Comment (superdump) [inline crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:36]: @IceSentry how is this?
```suggestion
    // Batches of fixed-size arrays of T are written to this buffer so that
    // each batch in a fixed-size array can be bound at a dynamic offset.
    uniforms: DynamicUniformBuffer<MaxCapacityArray<Vec<T>>>,
    // A batch of T are gathered into this `MaxCapacityArray` until it is full,
    // then it is written into the `DynamicUniformBuffer`, cleared, and new T
    // are gathered here, and so on for each batch.
    temp: MaxCapacityArray<Vec<T>>,
```

**Review by superdump — CHANGES_REQUESTED** [review]
Let's get this merged after this docs change, pending another approval.

File: crates/bevy_render/src/render_resource/batched_uniform_buffer.rs
Comment (konsolas) [inline crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:115]: As it stands, this code emits a divide even though it is only used with `dynamic_offset_alignment`, which is guaranteed to be a power of 2. 

Would suggest instead:
```rust
fn align_to_next(value: u64, alignment: u64) -> u64 {
    (value | (alignment - 1)) + 1
}
```

(or a similar name)

Comment (robtfm) [reply crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:115]: maybe `((value - 1) | (alignment - 1)) + 1`?

Comment (robtfm) [reply crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:115]: only works if we're sure the alignment is a power of 2. i think we are here but maybe we could use the log2 as the param to enforce it

Comment (konsolas) [reply crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:115]: > maybe `((value - 1) | (alignment - 1)) + 1`?

Yep you're right about this.

I think we can be very sure the alignment is a power of 2 (as it is an alignment after all) but the value comes from wgpu in this non-log2 form. How about:

```rust
fn align_to_next(value: u64, alignment: u64) -> u64 {
    debug_assert!(alignment & (alignment - 1) == 0);
    ((value - 1) | (alignment - 1)) + 1
}
```

**Review by robtfm — APPROVED** [review]

File: crates/bevy_render/src/render_resource/batched_uniform_buffer.rs
Comment (IceSentry) [reply crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:36]: LGTM

File: crates/bevy_render/src/render_resource/batched_uniform_buffer.rs
Comment (superdump) [reply crates/bevy_render/src/render_resource/batched_uniform_buffer.rs:120]: I'm fine with making this change.

Comment (superdump) [reply crates/bevy_render/src/render_resource/batched_uniform_buffer.rs]: I committed this with attribution to @konsolas and @robtfm.

**Review by superdump — APPROVED** [review]

**Review by robtfm — APPROVED** [review]

**Review by konsolas — APPROVED** [review]

**Pull request merged**
